### PR TITLE
service: Simplify logic of svc ID allocation

### PR DESCRIPTION
**Pull request by brb**
This PR changes the logic of the service ID allocation which allows to simplify a code of the callers.

Previously, if a requested service ID had been allocated to another service, the allocation request silently allocated a new service ID. This behavior required additional checks at the caller side to validate whether the allocated ID and the requested ID were the same. This could lead to a service ID leak if the caller forgot to release the ID if the IDs were not the same.

The change is to return an error if a requested ID is already taken. This allows us to get rid of some boilerplate code when creating a service.

Reviewable per commit.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9213)

### CODE REVIEW

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

File: daemon/k8s_watcher.go [hunk 1]
```diff
@@ -1256,25 +1256,23 @@ func (d *Daemon) addK8sSVCs(svcID k8s.ServiceID, svc *k8s.Service, endpoints *k8
...
-				}).Error("Error while getting a new service ID. Ignoring service...")
-				continue
-			}
-			scopedLog.WithFields(logrus.Fields{
-				logfields.ServiceName: fePortName,
-				logfields.ServiceID:   feAddrID.ID,
-				logfields.Object:      logfields.Repr(svc),
-			}).Debug("Got feAddr ID for service")
-			fePort.ID = loadbalancer.ServiceID(feAddrID.ID)
+		feAddr := loadbalancer.NewL3n4Addr(fePort.Protocol, svc.FrontendIP, fePort.Port)
```
Comment (aanm) [inline daemon/k8s_watcher.go]: what about when a service is added through the cilium-agent's CLI?

File: daemon/loadbalancer.go [hunk 2]
```diff
@@ -130,7 +140,8 @@ func (d *Daemon) svcAdd(
 	// of the backends when listing services. This step will go away once
 	// we start acquiring backend IDs in this module.
 	for i, be := range svc.BES {
-		id, err := service.LookupBackendID(be.L3n4Addr)
+		var id loadbalancer.BackendID
+		id, err = service.LookupBackendID(be.L3n4Addr)
 		if err != nil {
 			scopedLog.WithField(logfields.BackendName, be.L3n4Addr).WithError(err).
 				Warning("Unable to lookup backend ID")
```
Comment (aanm) [inline daemon/loadbalancer.go]: Why do we continue if err != nil? If err is not nil for index 0 but is nil for index 1, why is that ok?

**Review by aanm — CHANGES_REQUESTED** [review]

File: daemon/k8s_watcher.go (hunk 1 above)
Comment (brb) [reply daemon/k8s_watcher.go]: You can specify it via the `--id` param. E.g. `cilium service update ... --id 1234`.

File: daemon/loadbalancer.go (hunk 2 above)
Comment (brb) [reply daemon/loadbalancer.go]: This is unrelated to my PR, and this is not critical, as the retrieved IDs are used only for `cilium service list`.

Anyway, I have in my LB refactoring TODO to get rid of this.

Comment (aanm) [reply daemon/loadbalancer.go]: @brb it is related with your PR because the `err` in line 144 can be the same of line `99`

**Comment by brb** [comment]
test-me-please

File: daemon/loadbalancer.go (hunk 2 above)
Comment (brb) [reply daemon/loadbalancer.go]: Oops, nice catch. Fixed.

**Comment by brb** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Pull request merged**